IDE: add implementations Code Vision hint

Kobzol opened the pull request:
The `handleClick` implementation is copied from `ImplsNavigationGutterIconRenderer::navigateToItems`. It should probably be placed into some unified location, but I'm not sure where does it belong.

Relevant issue: https://github.com/intellij-rust/intellij-rust/issues/9427

changelog: Add support for showing implementations (inheritors) Code Vision. Note, these hints are available starting with 2022.3 and are currently disabled by default. You can turn it on via `Registry... | org.rust.code.vision.implementation`.

sirgl commented on src/223/main/kotlin/org/rust/ide/hints/codeVision/RsImplementationsCodeVisionProvider.kt:
```diff
@@ -0,0 +1,96 @@
+            is RsAbstractable -> element.searchForImplementations()
+            else -> return emptyList()
+        }
+        return query.findAll().toList()
```
If the name is too popular, this query is going to run forever and will be recalculated on every keystroke. For such names it is not worth to do the search, just skip. Also, you don't need to actually store elements, just the size is important.
Some kind of cache is also required, because inheritors depends only on the structure of declarations.

ortem replied:
@sirgl thank you for reviewing this!

Kobzol commented:
I added a fast path rejection and a cache, let me know what you think.

Undin commented on src/223/main/kotlin/org/rust/ide/hints/codeVision/RsImplementationsCodeVisionProvider.kt:
```diff
@@ -0,0 +1,120 @@
+private val RsElement.implementationCount: Int
+    get() = CachedValuesManager.getCachedValue(this, IMPL_CACHE_KEY) {
+        val usages = countImplementations(this)
+        CachedValueProvider.Result.create(usages, PsiModificationTracker.MODIFICATION_COUNT)
```
It doesn't seem it helps because cache with `PsiModificationTracker.MODIFICATION_COUNT` will drop after any code change

Kobzol replied:
Oh, this should have been the structure modification tracker, my bad. Thanks!

Undin commented on src/223/main/kotlin/org/rust/ide/hints/codeVision/RsImplementationsCodeVisionProvider.kt:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    override fun getHint(element: PsiElement, file: PsiFile): String? {
+        val rsElement = element as? RsElement ?: return null
+        if (rsElement !is RsTraitItem && rsElement !is RsAbstractable) return null
```
nit: if replace `val rsElement = element as? RsElement ?: return null` with `if (element !is RsElement) return null`, you will be able to use `element` and not to introduce a new variable with a new name

But in this case, you even don't need `val rsElement = element as? RsElement ?: return null` at all since you have more strict check at the next line.

Kobzol replied:
The following check doesn't help, because Kotlin isn't smart enough to smart cast to RsElement after checking that it isn't anstractable or trait (and there is no unifying interface for abstractables and trait I think).

Undin replied:
Oh, I see. But you still can avoid introducing redundant `rsElement` variable

Kobzol commented:
Rebased over `master` and moved stuff to the main source set. The code vision base class was still in the `223` source set on `master` for some reason. Was this intentional? I moved it to the main source set.

Kobzol commented:
Ah, it was intentional, now I understand why it's needed. I reverted the move.

Kobzol commented:
Rebased.

vlad20012 commented:
Since it is under a registry flag `org.rust.code.vision.implementation` I don't see any reasons not to merge it.

Later I'll measure highlighting performance impact and will enable it by default if it's fine.

Thanks!

bors r+
